### PR TITLE
[Feature] Checkbox component

### LINKED ISSUE #202: [Feature] Checkbox
**Is your feature request related to a problem? Please describe.**

Feature request: Checkbox. 

image
[source](https://material-ui.com/components/checkboxes/)

**Describe the solution you'd like**

- It muse have no dependencies.
- It must works with same style for all platforms.
- It must have a modifiable style.
- It muse be acceptable custom image.

**Describe alternatives you've considered**

**Additional context**

Refs
- https://material-ui.com/components/checkboxes/
- https://react-native-elements.github.io/react-native-elements/docs/checkbox.html

### PR DESCRIPTION
**Pull request by JinleeJeong**
## Description

CheckboxGroup: Multiple values ​​can be managed through the checkbox component.
Checkbox: Values ​​can be managed through checkbox.

| Checkbox | CheckboxGroup | CheckAll |   
|---------|-------------|------------|
| Checkbox | CheckboxGroup | CheckAll | 
## Related Issues

Resolve #202 

## Tests

I added the following tests:

Test file path: `main/__tests__/Checkbox.test.tsx`
Story: `stories/dooboo-ui/Checkbox.stories.tsx`
simple snap test of only Checkbox

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/dooboo-ui/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

File: main/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
+import React, { FC, useCallback, useContext, useEffect, useRef, useState } from 'react';
+import { GroupCheckboxContext } from './CheckboxGroup';
+import { TouchableHighlight } from 'react-native';
+import styled from 'styled-components/native';
+
+interface customStyle {
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: Interface should be started capital letter.

File: main/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  labelSize?: number;
+  labelColor?: string;
+  boxSize?: number;
+  boxColor?: string;
+  labelLeft?: boolean;
+}
+
+interface onChangeEvent {
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: Use Pascal Case.

File: main/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  checked = false,
+  defaultChecked = false,
+  indeterminate = false,
+  disabled = false,
+  onChange,
+  customStyle,
+
+}) => {
+  const groupCheckboxContext = useContext(GroupCheckboxContext);
+
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: It looks useless line break.

File: main/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+              source={{ uri: checkboxImg }}
+            />}
+            {
+              indeterminate && <Markerindeterminate
+                boxColor={disabled ? COLOR.LIGHTGRAY : customStyle?.boxColor}
+              />
+            }
+          </Marker>
+        </MarkerContainer>
+        <Label labelColor={labelColor} labelSize={customStyle?.labelSize} labelLeft={customStyle?.labelLeft}>
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: Multiline will be more readable.

File: main/Checkbox/Checkbox.tsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+          {label}
+        </Label>
+      </Container>
+    </TouchableHighlight>
+  );
+};
+
+interface ContainerProps {
+  labelLeft?: boolean;
+}
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx:108]: Line break.

File: main/Checkbox/CheckboxGroup.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+import styled from 'styled-components/native';
+
+interface customStyle {
+  labelSize?: number;
+  labelColor?: string;
+  boxSize?: number;
+  boxColor?: string;
+}
+
+interface onChangeEvent {
```
Comment (JeffGuKang) [inline main/Checkbox/CheckboxGroup.tsx]: Pascal Case.

File: main/Checkbox/CheckboxGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
+import React, { FC, useCallback, useEffect, useRef, useState } from 'react';
+import Checkbox from './Checkbox';
+import styled from 'styled-components/native';
+
+interface customStyle {
```
Comment (JeffGuKang) [inline main/Checkbox/CheckboxGroup.tsx]: Pascal Case.

File: main/Checkbox/Checkbox.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+  BLUE_SKY: '#1890FF',
+  LIGHTGRAY: '#E0E0E0',
+  GRAY59: '#969696',
+  BLACK: '#000000',
+};
+
+const Container = styled.View<ContainerProps>`
+ flex-direction: ${({ labelLeft }): string => labelLeft ? 'row-reverse' : 'row'};
+ justify-content: center;
+ align-items: center;
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: We are using 2 spaces. But it looks only one space.

File: main/Checkbox/Checkbox.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+`;
+
+const MarkerContainer = styled.View<MarkerContainerProps>`
+ padding: 1.5px;
+ width: ${({ boxSize }):number => boxSize || 20}px;
+ height: ${({ boxSize }):number => boxSize || 20}px;
+ background-color: ${({ boxColor }): string => boxColor || COLOR.BLUE_SKY};
```
Comment (JeffGuKang) [inline main/Checkbox/Checkbox.tsx]: 2 spaces.

Comment (JeffGuKang) [reply main/Checkbox/Checkbox.tsx]: And other styled components have a same space issue.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+      <Container>
+        <Title>with disabled props</Title>
+        <CheckboxGroup options={options} disabled onChange={onChange} />
+      </Container>
+
+    </ScrollContainer>
+  );
+}
+
+function CheckAll(): React.ReactElement {
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: How about `CheckAllExample`? Cause you are using checkAll for state inside.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+  const onCheckAllChange = (e) : void => {
+    setCheckedList(e.checked ? plainOptions : []);
+    setIndeterminate(false);
+    setCheckAll(e.checked);
+  };
+
+  return (
+    <ScrollContainer>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    { label: 'Orange', value: 'Orange', disabled: true },
+  ];
+
+  const onChange = (checkedValues): void => {
+    // eslint-disable-next-line no-console
+    console.log('checked = ', checkedValues);
+  };
+  return (
+    <ScrollContainer>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+      <Container>
+        <Title>plainOption</Title>
+        <CheckboxGroup options={plainOptions} onChange={onChange} />
+      </Container>
+      <Container>
+        <Title>Default</Title>
+        <CheckboxGroup options={options} onChange={onChange} />
+      </Container>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+      <Container>
+        <Title>option with disabled</Title>
+        <CheckboxGroup options={optionsWithDisabled} onChange={onChange} />
+      </Container>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+      <Container>
+        <Title>with disabled props</Title>
+        <CheckboxGroup options={options} disabled onChange={onChange} />
+      </Container>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

Comment (JeffGuKang) [reply stories/dooboo-ui/Checkbox.stories.tsx]: Line break is good way to divide parts. But it used without any rule here.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  const customStyle = {
+    labelLeft: true,
+  };
+  return (
+    <ScrollContainer>
+
+      <Container>
+        <Title>default</Title>
+        <Checkbox label="defaultChecked" defaultChecked ></Checkbox>
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line break.

File: stories/dooboo-ui/Checkbox.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,179 @@
+import { Checkbox, CheckboxGroup } from '../../main';
+import React, { ReactElement, useState } from 'react';
+
+import { ContainerDeco } from '../../storybook/decorators';
+
+import { storiesOf } from '@storybook/react-native';
+import styled from 'styled-components/native';
+
```
Comment (JeffGuKang) [inline stories/dooboo-ui/Checkbox.stories.tsx]: Useless line breaks.

**Review by JeffGuKang — CHANGES_REQUESTED** [review]
I am not sure `Context` is a good way to control `CheckboxGroup`. It could be improved later. 
Anyway `Checkbox` worked well when I tested. It will be merged after fixing some minor issues I reviewed.

File: main/Checkbox/Checkbox.tsx (hunk 8 above)
Comment (JinleeJeong) [reply main/Checkbox/Checkbox.tsx:138]: Thank you for the detailed review. Could you please explain a little more?

File: stories/dooboo-ui/Checkbox.stories.tsx (hunk 10 above)
Comment (JinleeJeong) [reply stories/dooboo-ui/Checkbox.stories.tsx]: Thank you for the detailed review. We have to work on this part. :)

Comment (p-iknow) [reply stories/dooboo-ui/Checkbox.stories.tsx]: I changed CheckAll to CheckAllExample. :)

**Comment by JinleeJeong** [comment]
We have to write the test code :) After writing, We will commit again. 
Thank you for the review!

**Comment by JeffGuKang** [comment]
> We have to write the test code :) After writing, We will commit again.
> Thank you for the review!

I added WIP to the PR's title temporarily.

**Comment by JeffGuKang** [comment]
The minor lint issues will be solved later after I add more lint rules. 
LGTM!

**Pull request merged**